optuna/storages/rdb/models.py

```diff
@@ -145,7 +146,11 @@ def where_study_id(cls, study_id, session):
 
 class TrialModel(BaseModel):
     __tablename__ = 'trials'
+    __table_args__ = (
+        Index('ix_study_trial_number', 'study_id', 'number'),
```

Comment:
Could you tell me why you use index instead of a unique constraint?
I found this commit (https://github.com/optuna/optuna/pull/939/commits/edc76350e8af23c1cb2b3d50bd46d8c5af81184d) that replaced the unique constraint with the index.

Currently, `number` is not used in any queries, and I don't think it is a typical usage of the index.
In addition, the Index class of SQLAlchemy is not necessarily unique (c.f., https://docs.sqlalchemy.org/en/13/core/constraints.html#sqlalchemy.schema.Index.params.unique).

Reply:
Yes. I tried to add unique constraints, but I removed unique constraints at e89580b0.

To add unique constraints, it requires table-level locking while inserting a `trials` record. 
But mechanisms of table-level locking heavily depend on RDBMS.

* SQLite3: To lock tables, we need to start a transaction with a `BEGIN EXCLUSIVE` statement.
    * https://www.sqlite.org/lang_transaction.html
* InnoDB(MySQL): MySQL uses row-level locking for InnoDB tables. So we need to call `LOCK TABLES ... WRITE` before inserting a request.
    * https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html
* MyISAM(MySQL): MySQL uses table-level locking for MyISAM tables. So it *doesn't* require to call `LOCK TABLES ... WRITE` statement.
    * https://dev.mysql.com/doc/refman/8.0/en/internal-locking.html

These differences make the code quite complex. For example, the following code is required for table-level locking on SQLite3. But it should be disabled when using MySQL. Furthermore, we need to call `LOCK TABLES` statement if the table is managed by InnoDB.

```python
# See https://docs.sqlalchemy.org/en/13/dialects/sqlite.html
@event.listens_for(engine, "begin")
def do_begin(conn):
    conn.execute("BEGIN EXCLUSIVE")
```

These are the reason why I avoid to set unique constraints.

---

> Currently, number is not used in any queries, and I don't think it is a typical usage of the index.

Yes, you are right. I once thought that there are some advantages to set a multiple-column index like the following discussion.

https://github.com/optuna/optuna/pull/631#issuecomment-545296339

But on second thought, multiple-column index `study_id, trial_id` is more useful. because we often query trials with study_id and trial_id. And the order of trial_id and number are the same.

Reply:
I agree that creating an index on the `number` column isn't really beneficial. As for introducing uniqueness constraints, if it's technically difficult, I think it's okay to leave it simple, trying best-effort to isolate the creation of trials (i.e. `create_new_trial`). Handling uniqueness conflicts isn't straight forward neither?

---

> But on second thought, multiple-column index study_id, trial_id is more useful

This sounds reasonable. I wonder what performance gains we could get. It might be worth discussing in a separate issue?

Reply:
> Handling uniqueness conflicts isn't straight forward neither?

It's easy to handle uniqueness because the current implementation is based on `trial_id` which is an AUTOINCREMENT field that is ensured to be monotonically increasing in the system level, not user's query level.

---

> I wonder what performance gains we could get. It might be worth discussing in a separate issue?

It's OK to discuss this matter in a separate issue. In my recognition, it can obviously avoid full-scan on some performance important API that are called multiple times like [here](https://github.com/optuna/optuna/blob/b60e4d4f5f563e6d88bb62ba79d1c50bbaee8b33/optuna/storages/rdb/models.py#L205-L207).

Reply:
> Handling uniqueness conflicts isn't straight forward neither?

Sorry, let me take this back.

Reply:
@c-bata Thank you for your explanation. I misunderstood that sqlalchemy abstracted  lock handling.

> the current implementation is based on trial_id which is an AUTOINCREMENT field that is ensured to be monotonically increasing in the system level, not user's query level.

Makes sense. The unique constraint for trial number is not required because the uniqueness of trial number is guaranteed by the logic to generate the number.

> It's OK to discuss this matter in a separate issue.

In terms of the discussion on index, I also agree with you and @hvy. I'd like to measure the improvement if we introduce the index because it accelerates the queries but requires additional computation for insertion.

(This is not directly related to this issue.)
We can see the use of UniqueConstraint in other part of `oputna/storages/rdb/models.py`. Do we need to revise unique constraints as well?

```python
class TrialSystemAttributeModel(BaseModel):
    __tablename__ = 'trial_system_attributes'
    __table_args__ = (UniqueConstraint('trial_id', 'key'), )  # type: Any
```

Reply:
> In terms of the discussion on index, I also agree with you and @hvy. I'd like to measure the improvement if we introduce the index because it accelerates the queries but requires additional computation for insertion.

Okay. I'll revert the changes for the index here.

> We can see the use of UniqueConstraint in other part of oputna/storages/rdb/models.py. Do we need to revise unique constraints as well?

I've just checked the constraints of system_attrs and user_attrs with following snippet.

```python
import optuna
from concurrent.futures.thread import ThreadPoolExecutor


def main():
    optuna.logging.set_verbosity(optuna.logging.DEBUG)
    study = optuna.create_study(
        storage='sqlite:///db.sqlite3',
    )

    with ThreadPoolExecutor(10) as pool:
        pool.map(
            lambda x: study.set_user_attr('foo', x),
            range(20),
            timeout=20,
        )
    print("Exit")


if __name__ == '__main__':
    main()
```

```
$ python examples/concurrent_insert.py
[I 2020-03-05 16:53:27,602] A new study created with name: no-name-35958551-e7f2-41e0-988c-1d5567e8da0e
[D 2020-03-05 16:53:27,637] Ignoring IntegrityError('(sqlite3.IntegrityError) UNIQUE constraint failed: study_user_attributes.study_id, study_user_attributes.key'). This happens due to a timing issue among threads/processes/nodes. Another one might have committed a record with the same key(s).
Exit
```

It seems to just rollback by _commit_with_integrity_check if keys are conflicted. So there are no serious problems with the current usages of UniqueConstraints.

https://github.com/optuna/optuna/blob/ac39fee7b0a56764e518d5569f5b97b3b748446e/optuna/storages/rdb/storage.py#L889-L901